File: 0016-3sum-closest/0016-3sum-closest.cpp
/*
[-4,-1,1,2]

*/
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(),nums.end());
        int n = nums.size();
        int differ = INT_MAX;
        int ans = INT_MAX;
        for(int i = 0; i<nums.size(); i++){
            int curr = nums[i];
            int l = i +1, r= n-1;
            while(l<r){
                int sum = curr+nums[l] + nums[r];
                if(sum > target){
                    if(differ > abs(target-sum)){
                        ans = sum;
                        differ = abs(target-sum);
                    }
                    r--;
                }else if(sum < target){
                    if(differ > abs(target-sum)){
                        ans = sum;
                        differ = abs(target-sum);
                    }
                    l++;
                }else{
                    return target;
                }
            }
        }
    return ans;
    }
};

File: 0016-3sum-closest/0016-3sum-closest.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
    let [differ,ans] = [Number.MAX_VALUE,Number.MAX_VALUE];
    nums.sort((a,b)=>a-b);
    const n = nums.length;
    for(let i = 0; i<n; i++){
        let [l,r] = [i+1,n-1];
        while(l<r){
            const sum = nums[i] + nums[l] + nums[r];
            if(sum > target){
                if(differ > Math.abs(target-sum)){
                    ans = sum;
                    differ = Math.abs(target-sum);
                } 
                r--;
            }else if(sum < target){
                if(differ > Math.abs(target-sum)){
                    ans = sum;
                    differ = Math.abs(target-sum);
                }
                l++;
            }else{
                return sum;
            }
        }
    }
    return ans;
};